walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,68 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.List;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
 
 public final class FindMeetingQuery {
+
+  /** 
+   * Return list of unavialable times to schedule the meeting. 
+   * List of TimeRanges is in sorted order from earliest to latest start time.
+   */
+  private List<TimeRange> getUnavailableTimes(Collection<Event> events, Collection<String> attendees) {
+    HashSet<TimeRange> unavailableTimesSet = new HashSet<>();
+    for (Event e : events) {
+      Collection<String> eventAttendees = e.getAttendees();
+      if (!Collections.disjoint(eventAttendees, attendees)) {
+        unavailableTimesSet.add(e.getWhen());
+      }
+    }
+
+    List<TimeRange> unavailableTimesList = new ArrayList<>(unavailableTimesSet);
+    Collections.sort(unavailableTimesList, TimeRange.ORDER_BY_START);
+    return unavailableTimesList;
+  }
+
+  /**
+   * Return list of possible meeting times based on the list of unavailable times and the duration of the desired meeting. 
+   */
+  private List<TimeRange> getMeetingTimes(List<TimeRange> unavailableTimes, long duration) {
+    List<TimeRange> meetingTimes = new ArrayList<>();
+    int firstAvail = TimeRange.START_OF_DAY;
+
+    for (TimeRange t : unavailableTimes) {
+      int start = t.start();
+      int end = t.end();
+      if (start - firstAvail >= duration) {
+        meetingTimes.add(TimeRange.fromStartEnd(firstAvail, start, false));
+      }
+      firstAvail = Math.max(end, firstAvail);
+    }
+    if (TimeRange.END_OF_DAY - firstAvail >= duration) {
+      meetingTimes.add(TimeRange.fromStartEnd(firstAvail, TimeRange.END_OF_DAY, true));
+    }
+
+    return meetingTimes;
+  }
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
```

Comment:
I'm reading this function as returning meeting times for all attendees if possible, otherwise it returns meeting times without optional attendees. I would love a description for this query function.